### ISSUE TITLE
Changing /home/$NB_USER for non-privileged containers.

### ISSUE THREAD

**Issue by DeepCowProductions** [jupyter/docker-stacks#1923]
### What docker image(s) is this feature applicable to?

base-notebook

### What changes are you proposing?

Add group write permissions for group "root" to /home inside the container, 
and add support for the creation of /home/$NB_USER for non-privileged containers.

### How does this affect the user?

This would allow a relatively easy way to create a new home folder for users/ids other than jovyan.

The current workaround involves either one or more of these options:
- mounting the "correct home" from outside into the container. E.g:  `docker run -e NB_USER="myuser" -v <myuser_home_host> /home/myuser`
- cd -ing into the new home and exporting $HOME to the correct path
- retrofitting the entry in /etc/passwd through post startup hooks.

It would be nice to have this feature in the base image.

### Anything else?

**Example implementation**
- Insert this into the base Dockerfile:
`chmod g+rwx /home`
- Set the home inside /etc/passwd:
`echo "${NB_USER}:x:$(id -u):$(id -g):,,,:/home/${NB_USER}:/bin/bash" >> /tmp/passwd`
- Replace the warning in start.sh Line 231 with the home folder creation, that already exists for the privileged container execution.

Since a single container is mostly run by a single user and the modifications to /home are not persistent (unless mounted), this is unlikely to have any negative side effects  (at least I'm not aware of them)

**Comment by mathbunnyru** [comment]
Unfortunately, I don't have enough confidence in linux user/group management to say if this is gonna be a good change or it might create issues for some users.

From the maintainer point of view, I think our `start.sh` is already way too complicated (at least half of our current issues are related to this script). If we want to add a new feature to it, it should be well justified and well tested.

**Comment by DeepCowProductions** [comment]
Over the next semester at our institution, we are going to test the workaround with `chmod g+rwx /home` in the Dockerfile and startup hooks to rebind the home folder. I am happy to report whether we run into any issues along the way. 
To add some background: we need to change the user inside the container to monitor storage consumption of our students.

The way I see it, there is no simple way of getting rid of `start.sh` without losing the features and the flexibility that it provides.

Regarding linux user/group management, the choice of `/home` for the user homes is a rather arbitrary one, and read and write permissions in this context is only relevant for data privacy on multi-user systems.

**Comment by DeepCowProductions** [comment]
Okay, I can provide some more information now.

TLDR: It works, but whether theese changes should make it into the base containers, I am not sure....
Rebinding the user home to `/home/$NB_USER` and launching with a user other then jovyan 1000:100 requires more changes outside of this repository.
Its also more hassle than its worth.


I will leave a story here for anyone that requires such a rebind for running a Jupterhub on Kubernetes, where the files created by the users on the shared storage need to be associated to the individual user via user id and group. This a the use case at our institution.
The following also works analogously for pure docker or docker swarm, in case someone wants to start a container as "myuser" (may or may not be a be system user with id 5000)
`docker run --user 5000:100 --group-add root --group-add 1000 -e NB_USER="myuser" <image>`


In the case of Kubernetes an initContainer with either root or extended [cabalilties](https://manpages.ubuntu.com/manpages/xenial/en/man7/capabilities.7.html) has to be created to assign the correct file permissions to the mounted home directory. 

An Image needs something like this to do this without getting run as root:
```Dockerfile
USER root

RUN setcap cap_chown,cap_fowner,cap_dac_read_search,cap_dac_override,cap_fsetid,cap_setuid,cap_setgid+ep /bin/chown
RUN setcap cap_chown,cap_fowner,cap_dac_read_search,cap_dac_override,cap_fsetid,cap_setuid,cap_setgid+ep /bin/chmod
```

Let's  say your user has some id and group, that you can also set in some kubespawner hooks for the pod context dynamically:
```yaml
spec:
    securityContext:
        fsGroup: 1001
        runAsUser: 1001
        runAsGroup: 1001
```
Then you need to make sure that either `/home/$NB_USER` or `/home/jovyian` is readable/writable. The setting "fsGroupChangePolicy" and "fsGroup" were not working for us to ensure the latter.

An initContainer for the singleuser config could look like this:
```yaml
singleuser:
    initContainers:
        -   name: permissionfix
        securityContext: 
            capabilities: 
            add: ["CHOWN", "FOWNER", "dac_read_search", "dac_override", "fsetid", "setuid", "setgid"]
        image: ghcr.io/fhswf/jupyterhub-k8s/multitool:sha-399fe69 # image with extended syscab on chown and chmod
        command: ["/bin/sh","-c"]
        args: ["id; find /userdata -not -user 1001 | xargs -I % echo 'chown 1001 \"%\"'; echo 'done'"] # find all with wrong permissions
        volumeMounts:
        -   name: volume-{username}  #  home mount for the user
            mountPath: "/userdata"
```
Because Jupyterhub can theoretically launch more than just the image from jupyter/docker-stacks, we want to make sure that home rebinds and changes to the user via securityContext work for other images as well. Such a hook can help with that:

```yaml
singleuser:
    lifecycleHooks:
        postStart:
        exec:
            command: [ "/bin/bash", "-c", "echo 'checking HOME' &&
            if [[ ${HOME} != \"/home/${NB_USER}\" ]]; then
                export HOME=/home/${NB_USER};
            fi &&
            if [[ ${PWD}/ == \"/home/jovyan/\"* ]]; then
                new_wd=\"/home/${NB_USER}/${PWD:13}\";
                echo \"Changing working directory to ${new_wd}\";
                cd \"${new_wd}\";
            fi &&
            echo 'done'"]
```

However, it is better to do this in the image itself. The following bash script does this. This would also be the part that could be integrated into docker-stacks...
```bash
#!/bin/bash
set -e

# works for command: docker run --user 5000:100 --group-add root --group-add 1000 -e NB_USER="myuser" ...
if [[ -w /etc/passwd ]] && [[ -w /home ]]; then
    if [[ "${NB_USER}" != "jovyan" ]]; then
         _log "creating enw home for user ${NB_USER}"
        mkdir -p /home/${NB_USER};
        head -n -1 /etc/passwd > /tmp/passwd;  # remove last line, it would be better to replace jovyian line with a regex ....
        echo "creating passwd entry ${NB_USER}:x:$(id -u):$(id -g):,,,:/home/${NB_USER}:/bin/bash";
        echo "${NB_USER}:x:$(id -u):$(id -g):,,,:/home/${NB_USER}:/bin/bash" >> /tmp/passwd;
        cat /tmp/passwd > /etc/passwd;
        rm /tmp/passwd;
        if  [[ $HOME !=  "/home/${NB_USER}" ]]; then
            export HOME=/home/${NB_USER}
        fi
        if [[ ${PWD}/ == "/home/jovyan/"* ]]; then
            new_wd="/home/${NB_USER}/${PWD:13}"
            echo "Changing working directory to ${new_wd}"
            cd "${new_wd}"
        fi
        # snippets taken from the docker-stacks start.sh that do not get executed atm.
        # note that file system permissions for /home/${NB_USER} have to be checked/corercted with a privilieged initContainer for Kubernetes
        export XDG_CACHE_HOME="/home/${NB_USER}/.cache"
    fi
else
    echo "WARNING: unable to fix missing /etc/passwd entry because we don't have write permission. Try setting gid=0 with \"--user=$(id -u):0\"."
fi
```
It is copied to the before-notebook hooks location. So you have to extend the Dockerfiles from this repo with something like this:
```Dockerfile
ARG JUPYTER_VARIANT
FROM jupyter/${JUPYTER_VARIANT}-notebook
USER root
# make /home group writable in case we want to move /home/$NB_USER
RUN chmod g+rwx /home
# copy start scripts
COPY rebind_home.sh /usr/local/bin/before-notebook.d/01_rebind_home.sh
RUN  chmod +x /usr/local/bin/before-notebook.d/01_rebind_home.sh
USER $NB_UID
``` 

All of this requires, that a user (coming from some kind of identity provider) has a user id attached to it and that whatever spawner you use sets that user id during spawn in the places where it is required.
```python 
jupyterhub_config.py:

def get_uid(spawner):
    # called after getprofile when user wants to spawn a selected profile
    spawner.log.info("get uid/gid for " + str(spawner._expand_all("{username}")))
    # wget/curl or from auth_state => new_id = ... 
    spawner.uid = new_id
    spawner.environment["NB_USER"] = spawner._expand_all("{username}")
    return new_id
c.KubeSpawner.uid = get_uid

def get_pod_security_context(spawner):
    pod_security_context = get_config("custom.podSecurityContext")
    uid = spawner.uid
    pod_security_context["runAsUser"] = uid
    return pod_security_context
c.KubeSpawner.pod_security_context = get_pod_security_context
```
There could an easier way to achive similar, but when you want to go full cricle and have your users and ids consistent everywhere, this is one way.

**Comment by mathbunnyru** [comment]
@DeepCowProductions thank you for your thorough explanation.

> TLDR: It works, but whether theese changes should make it into the base containers, I am not sure....

I think this change is a bit too specific for your use case, so it would be better not to include it.
Otherwise, we'll have to have a lot of discussions about whether this should be included, lots of tests, and so on.
And even after that, `chmod g+rwx /home` might be breaking for some users.

I suggest you take a look at <https://github.com/jupyter/cookiecutter-docker-stacks> and create your image based on one of our images. This will allow you to easily have scheduled builds and workflow to test your image.
As far as I understand, you do it just by adding a `before-notebook.d` hook and a few lines of Dockerfile, which is quite nice (and it is unlikely to break).
